### PR TITLE
NAS-116409 / None / Add basic capacity management card and ix-dataset-node component

### PR DESCRIPTION
**Pull request by denysbutenko**
image

### CODE REVIEW

File: src/app/pages/datasets/components/dataset-management/dataset-management.component.ts [hunk 1]
```diff
@@ -18,10 +16,12 @@ import { DatasetNode } from './dataset-node.interface';
...
   selectedDataset: Dataset; // Dataset to be passed as input for card components
-  dataSource: IxNestedTreeDataSource<DatasetNode>;
-  treeControl = new NestedTreeControl<IxTreeNode<Dataset>>((node) => node.children);
-  readonly trackByFn: TrackByFunction<IxTreeNode<Dataset>> = (_, node) => node.label;
-  readonly hasNestedChild = (_: number, nodeData: DatasetNode): boolean => !!nodeData.children?.length;
+  dataSource: IxNestedTreeDataSource<Dataset>;
+  treeControl = new NestedTreeControl<Dataset, string>((dataset) => dataset.children, {
+    trackBy: (dataset) => dataset.id,
+  });
+  readonly trackByFn: TrackByFunction<Dataset> = (_, node) => node.id;
```
Comment (RehanY147) [inline src/app/pages/datasets/components/dataset-management/dataset-management.component.ts:23]: Might as well replace `node` with `dataset` here too

**Review by RehanY147 — CHANGES_REQUESTED** [review]

File: src/app/pages/datasets/components/dataset-management/dataset-management.component.html [hunk 2]
```diff
@@ -60,41 +23,41 @@ <h3>{{ 'Loading...' | translate }}</h3>
...
-              [class.selected]="node.item.name === selectedDataset?.name"
+              (click)="onDatasetSelected(dataset)"
+              [class.selected]="dataset.name === selectedDataset?.name"
             >
               <button
                 mat-icon-button
                 ixTreeNodeToggle
                 (click)="$event.preventDefault()"
-                [attr.aria-label]="'toggle ' + node.label"
+                [attr.aria-label]="'toggle ' + dataset.name"
```
Comment (undsoft) [inline src/app/pages/datasets/components/dataset-management/dataset-management.component.html:50]: Should be translated.

File: src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    this.subscription = this.ws.call('pool.dataset.query', [[['id', '=', this.dataset.id]]]).pipe(
+      take(1),
+      map((datasets) => datasets[0]),
+      untilDestroyed(this),
+    ).subscribe(
+      (dataset) => {
+        this.dataset = dataset;
+        if (!dataset.encrypted) {
+          combineLatest([
```
Comment (undsoft) [inline src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.ts:53]: forkJoin without `take(1)` should be better suited here.

File: src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  ) {}
+
+  ngOnChanges(): void {
+    this.loading = true;
+    this.cdr.markForCheck();
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+    this.subscription = this.ws.call('pool.dataset.query', [[['id', '=', this.dataset.id]]]).pipe(
+      take(1),
```
Comment (undsoft) [inline src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.ts:46]: I don't think take(1) is necessary.

File: src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+  constructor(
+    private ws: WebSocketService,
+    private cdr: ChangeDetectorRef,
+    private modalService: ModalService,
+    private translate: TranslateService,
+    private mdDialog: MatDialog,
+  ) {}
+
+  ngOnChanges(): void {
```
Comment (undsoft) [inline src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.ts:39]: Would be good to split this into a method or two.

File: src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.html [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      <div class="legend-placeholder" [class.loading]="loading"></div>
+    </div>
+  </mat-card-content>
+  <mat-card-actions fxLayout="column" *ngIf="!dataset.encrypted">
+    <div class="row">
+      <span class="header">{{ 'User Quotas' | translate }}: </span>
+      <div fxFlex fxLayout="row" fxLayoutAlign="space-between center">
+        <span class="loader" *ngIf="loading"></span>
+        <span *ngIf="!loading">{{ 'Quotas set for {n} users' | translate: { n: userQuotas } }}</span>
+        <a [routerLink]="['/storage/user-quotas', dataset.id]">{{ 'Manage User Quotas' | translate }}</a>
```
Comment (undsoft) [inline src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.html:24]: Remember to copy quota management to new module in this or future PRs.

File: src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      .afterClosed()
+      .pipe(untilDestroyed(this))
+      .subscribe((shouldRefresh) => {
+        if (shouldRefresh) {
+          this.ngOnChanges();
+        }
+      });
+  }
+
+  editDataset(): void {
```
[the remark this answers is not in the document]
Comment (denysbutenko) [reply src/app/pages/datasets/components/dataset-capacity-management-card/dataset-capacity-management-card.component.ts:89]: What would be the preferred way to do that. Separate component or tweak an existing form to display quotas fields only?

**Review by undsoft — APPROVED** [review]
👍🏻

**Review by RehanY147 — APPROVED** [review]
Looks good now. I've created a ticket for the chart here [NAS-116788](https://jira.ixsystems.com/browse/NAS-116788). Once designs are finalized. When ready the ticket will be available in the queue.

**Pull request merged**